### PR TITLE
rework c function pointers

### PR DESCRIPTION
**Pull request by t-bltg**
Fix https://github.com/JuliaParallel/MPI.jl/issues/688.

~~Not yet ready for review ;)~~

### CODE REVIEW

File: gen/src/MPIgenerator.jl
Comment (t-bltg) [inline gen/src/MPIgenerator.jl:27]: @Gnimuc, could you advise how to replace some node arguments ?

File: gen/src/MPIgenerator.jl
Comment (giordano) [inline gen/src/MPIgenerator.jl:18]: This isn't the only one wrong, all pointers which aren't buffers should be changed

Comment (t-bltg) [reply gen/src/MPIgenerator.jl:18]: Yes, I know, this is a draft.
I'm only opening this PR now to discuss Clang internals with @Gnimuc.

**Comment by Gnimuc** [comment]
Is it still a problem after merging https://github.com/JuliaInterop/Clang.jl/pull/406 ?

**Comment by t-bltg** [comment]
https://github.com/JuliaInterop/Clang.jl/pull/406 is only for fixing debugging and printing stuff.

The question I want to ask here is: "is it possible to modify node function arguments" ?

Or do we have to modify the underlying node exprs generated (this seems unfeasible, since information is lost on the way ...) ?

**Comment by Gnimuc** [comment]
> (this seems unfeasible, since information is lost on the way ...)

The only elegant way is to overload [`_get_func_arg `](https://github.com/JuliaInterop/Clang.jl/blob/master/src/generator/codegen.jl#L23).

**Comment by t-bltg** [comment]
This was fun actually.
Monkey patching `_get_func_arg` is horrible, but it works ...

**Comment by Gnimuc** [comment]
I'm trying to understand the root cause of #688. To me, it seems like Clang.jl doesn't handle `typedef function prototype` well for MPI's special usage.

**Comment by t-bltg** [comment]
By the way, is the `output_ignorelist` option from `generator.toml` broken ?

It doesn't filter the function definitions anymore ...

**Comment by Gnimuc** [comment]
Only on the master branch? This could be related: https://github.com/JuliaInterop/Clang.jl/pull/403

**Comment by giordano** [comment]
Worked for me in #692

**Comment by t-bltg** [comment]
> Only on the master branch?

Ha ok, we have to adjust the generator regexes.

**Comment by t-bltg** [comment]
> I'm trying to understand the root cause of https://github.com/JuliaParallel/MPI.jl/issues/688. To me, it seems like Clang.jl doesn't handle typedef function prototype well for MPI's special usage.

Right, I think we'd better fix this:
The declaration in `mpi.h` is:
```c
typedef int (MPI_Type_copy_attr_function)(MPI_Datatype, int, void *, void *, 
					  void *, int *);
```

**Comment by t-bltg** [comment]
All-right, so with the patch fixing function pointer (https://github.com/JuliaInterop/Clang.jl/pull/408), the local MPI test are green on linux.

**Comment by t-bltg** [comment]
So, with the `Clang.jl` function pointer patch in https://github.com/JuliaInterop/Clang.jl/pull/408, I think I got this right now.

I also [got confused with c function pointers](https://github.com/JuliaParallel/MPI.jl/issues/688#issuecomment-1353919247) ...

File: gen/src/MPIgenerator.jl
Comment (giordano) [inline gen/src/MPIgenerator.jl:2]: We only need to update `Clang.jl`, with no other changes to the script?

Comment (t-bltg) [reply gen/src/MPIgenerator.jl:2]: Yeah, I can remove this now, since this was used at early stage when monkey patching.

**Comment by t-bltg** [comment]
I also need to add a test for https://github.com/JuliaParallel/MPI.jl/issues/688.

**Comment by giordano** [comment]
> I also need to add a test for https://github.com/JuliaParallel/MPI.jl/issues/688.

Revert #689 :stuck_out_tongue:

**Comment by giordano** [comment]
> @giordano, what is in your test.jl from https://github.com/JuliaParallel/MPI.jl/issues/688 ?

https://github.com/JuliaParallel/MPI.jl/issues/684#issuecomment-1341698748, which is simply extracted from https://github.com/JuliaParallel/MPI.jl/blob/7da078cf179a4d39e555713acaa74c84dbbf6729/test/test_allgather.jl#L3-L13 and before #689 that was running on `Char` as well.  Also, `MPIDatatype` should be moved to the tests.

File: test/test_issues.jl
Comment (giordano) [inline test/test_issues.jl:5]: This is redundant if you restore `Char` in `MPIDatatype` (which I this point I believe is the right thing to do), the same code is executed in `test_allgather.jl` already.

Comment (t-bltg) [reply test/test_issues.jl]: I prefer to be explicit about it, but your call. Removed.

**Comment by giordano** [comment]
I can confirm the test script works with this branch of MPI.jl on Ookami with MVAPICH:
```console
[mosgiordano@fj004 mvapich]$ srun -n 2 julia --project test.jl 
(rank, C) = (0, ['\x01', '\x02'])
(rank, C) = (1, ['\x01', '\x02'])
[mosgiordano@fj004 mvapich]$ srun -n 4 julia --project test.jl 
(rank, C) = (1, ['\x01', '\x02', '\x03', '\x04'])
(rank, C) = (2, ['\x01', '\x02', '\x03', '\x04'])
(rank, C) = (3, ['\x01', '\x02', '\x03', '\x04'])
(rank, C) = (0, ['\x01', '\x02', '\x03', '\x04'])
[mosgiordano@fj004 mvapich]$ srun -n 8 julia --project test.jl 
(rank, C) = (3, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
(rank, C) = (2, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
(rank, C) = (5, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
(rank, C) = (6, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
(rank, C) = (7, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
(rank, C) = (1, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
(rank, C) = (0, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
(rank, C) = (4, ['\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\a', '\b'])
```

:rocket:

**Comment by t-bltg** [comment]
> I can confirm the test script works with this branch of MPI.jl on Ookami with MVAPICH:

So
```julia
julia> MPI.get_name(MPI.Datatype(Char))
""
```
remains expected ?

File: test/common.jl
Comment (giordano) [inline test/common.jl:26]: Missing newline
```suggestion
                            Float32, Float64, ComplexF32, ComplexF64} |> Base.uniontypes

```
Maybe also add a comment to explain these are the types we want to test with MPI and warn to not remove any of them? :smile:

**Comment by giordano** [comment]
> So
> ```julia
> julia> MPI.get_name(MPI.Datatype(Char))
> ""
> ```
> remains expected ?

Sadly yes:

```
julia> using MPI

julia> MPI.Init(;threadlevel=:single)
MPI.ThreadLevel(0)

julia> MPI.get_name(MPI.Datatype(Char))
""
```

I get the same with `MPICH_jll` (and MVAPICH has the MPICH ABI, so this shouldn't be surprising).

I don't know if there is a way to query the "MPI equivalent type" for `Char`.

File: test/common.jl
Comment (t-bltg) [reply test/common.jl]: Done. Also, I don't think we need `Base.uniontypes` right ?
IIUC, listing the types in a vector should be sufficient.

**Comment by t-bltg** [comment]
Let's reserve that fix for another PR.

File: test/common.jl
Comment (giordano) [reply test/common.jl]: Yeah, can also just be a vector.  In #689 suggested this is just a leftover from the past, so we can change it for whatever we need in the tests, which is always an iterable from what I can tell.

Comment (simonbyrne) [reply test/common.jl]: It might be worth keeping some type alias for native MPI types (which would exclude Char)

Comment (t-bltg) [reply test/common.jl]: Sorry I don't get this could you clarify ?

**Comment by giordano** [comment]
I spoke too soon, I hadn't closed the julia session:
```
julia> using MPI

julia> MPI.Init(;threadlevel=:single)
MPI.ThreadLevel(0)

julia> MPI.Datatype(Char)
MPI.Datatype(Char): 

julia> MPI.Finalize()

[813476] signal (11.1): Segmentation fault
in expression starting at REPL[4]:1
MPIR_Call_attr_delete at /lustre/software/mvapich2/gcc11/2.3.6/lib/libmpi.so (unknown line)
MPIR_Attr_delete_list at /lustre/software/mvapich2/gcc11/2.3.6/lib/libmpi.so (unknown line)
Allocations: 644473 (Pool: 643990; Big: 483); GC: 1
Segmentation fault (core dumped)
[mosgiordano@fj-debug2 mvapich]$
```
As I mentioned in https://github.com/JuliaParallel/MPI.jl/issues/688#issuecomment-1353846363, lots of segfaults happen at `MPI_Finalize`-time.

Also, this PR does ***not*** solve #684, there's a load of other segmentation faults I get there, with both MVAPICH2 and Open MPI, this is just the first one I got with MVAPICH.

**Comment by t-bltg** [comment]
> Also, this PR does not solve https://github.com/JuliaParallel/MPI.jl/issues/684, there's a load of other segmentation faults I get there

Ok, I will remove the `fix ...`.

**Comment by t-bltg** [comment]
By the way, buildkite should have a some timeouts, there are jobs running for a week now:
https://buildkite.com/julialang/mpi-dot-jl/builds/999
https://buildkite.com/julialang/mpi-dot-jl/builds/1001

File: src/buffers.jl
Comment (simonbyrne) [inline src/buffers.jl:7]: Can we keep this for now? These are the "native" MPI types (i.e. the ones that exist out of the box)

**Comment by simonbyrne** [comment]
> By the way, buildkite should have some timeouts, there are jobs running for a week now: https://buildkite.com/julialang/mpi-dot-jl/builds/999 https://buildkite.com/julialang/mpi-dot-jl/builds/1001

Not currently possible, unfortunately:
https://forum.buildkite.community/t/timeout-waiting-for-agent/188/4

File: src/buffers.jl
Comment (t-bltg) [reply src/buffers.jl:7]: Sure.

**Review by simonbyrne — APPROVED** [review]

**Comment by simonbyrne** [comment]
Ah, sorry i merged too early.

@giordano doesn't this include the change from https://github.com/JuliaParallel/MPI.jl/issues/688#issuecomment-1353923564?

**Comment by giordano** [comment]
I think it does: https://github.com/JuliaParallel/MPI.jl/pull/693/files#diff-b01acb82d6f2ed53e2d5907924a42adb7162f0e91fbfe6beb70aaaae0d0e2c36R162 ?

**Comment by simonbyrne** [comment]
Great!

**Pull request merged**
